### PR TITLE
ASV: benchmark for DataFrame.Update

### PR DESCRIPTION
**Pull request by aureliobarbosa**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

@rhshadrach and @datapythonista 

Benchmarks for the DataFrame.Update method are implemented, as suggested on PR #57637 by @datapythonista and based on the manual tests done by @rhshadrach on #55634. By running ASV locally we measure the change in the performance due to the fix on #57637, as indicated below:

| Change   | Main (2.2.1) [8da8b544]   | PR (3.0) [93ea131f]  |   Ratio | Benchmark (Parameter)                                   |
|----------|----------------------------|----------------------------------------|---------|---------------------------------------------------------|
| -        | 1.13±0.02ms                | 887±10μs                               |    0.78 | frame_methods.Update.time_to_update_random_indices      |
| -        | 1.19±0.02ms                | 840±9μs                                |    0.71 | frame_methods.Update.time_to_update_small_frame_big_arg |
| -        | 1.27±0.01ms                | 859±9μs                                |    0.67 | frame_methods.Update.time_to_update_big_frame_small_arg |

Regards

### CODE REVIEW

**Review by rhshadrach — CHANGES_REQUESTED** [review]
Thanks for the PR! This looks good, except the runtime for the benchmark is a bit low (can make it hard to differentiate regressions vs noise). Can you increase the parameters to get runtime up to ~10ms.

**Comment by aureliobarbosa** [comment]
> Can you increase the parameters to get runtime up to ~10ms.

@rhshadrach Think I am done here.

Data was increased twofold, from 10**4 to 10**6 points. Each test is now taking ~4ms. Didn't check the scaling, but it seems to be enough because the data size is similar to other benchmarks in the `frame_methods.py`. If needed, I can increase the data again. 

It would be interesting to have some guidance on the best data size for a useful test, but that would be a whole work on itself.

File: asv_bench/benchmarks/frame_methods.py [hunk 1]
```diff
@@ -862,4 +862,28 @@ def time_last_valid_index(self, dtype):
...
+        self.df = DataFrame(rng.uniform(size=(100_000, 10)))
+
+        idx = rng.choice(range(100_000), size=100_000, replace=False)
+        self.df_random = DataFrame(self.df, index=idx)
+
+        idx = rng.choice(range(100_000), size=10_000, replace=False)
+        cols = rng.choice(range(10), size=2, replace=False)
+        self.df_sample = DataFrame(
+            rng.uniform(size=(10_000, 2)), index=idx, columns=cols
+        )
```
Comment (rhshadrach) [inline asv_bench/benchmarks/frame_methods.py:877]:
```suggestion
        rng = np.random.default_rng()
        self.df = DataFrame(rng.uniform(size=(1_000_000, 10)))

        idx = rng.choice(range(10_000_000), size=1_000_000, replace=False)
        self.df_random = DataFrame(self.df, index=idx)

        idx = rng.choice(range(1_000_000), size=100_000, replace=False)
        cols = rng.choice(range(10), size=2, replace=False)
        self.df_sample = DataFrame(
            rng.uniform(size=(100_000, 2)), index=idx, columns=cols
        )
```

Can you go one more order of magnitude larger. You increased by two orders of magnitude and saw very little change in runtime - meaning that you were mostly measuring overhead. With the above changes, I'm seeing:

```
%timeit df.update(df_sample)
%timeit df_random.update(df_sample)
%timeit df_sample.update(df)
34.7 ms ± 324 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
17.3 ms ± 158 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
19.4 ms ± 123 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

which is more in the realm of what we want for ASVs.

**Review by rhshadrach — CHANGES_REQUESTED** [review]

File: asv_bench/benchmarks/frame_methods.py (hunk 1 above)
Comment (aureliobarbosa) [reply asv_bench/benchmarks/frame_methods.py]: Done.

Just for consistency, the benchmarks comparing the last version of DataFrame.Update() and the current one. At some point there is some performance crossover, for a smaller dataframe, it seems to be a win-win. But for larger, there is performance decrease in some cases. 

| Change   | Before [4fe49b16] <main>   | After [93ea131f] <fix_gh55509_v2~14>   |   Ratio | Benchmark (Parameter)                                   |
|----------|----------------------------|----------------------------------------|---------|---------------------------------------------------------|
| +        | 40.4±1ms                   | 50.7±0.9ms                             |    1.25 | frame_methods.Update.time_to_update_big_frame_small_arg |
| -        | 68.1±0.8ms                 | 35.9±0.9ms                             |    0.53 | frame_methods.Update.time_to_update_random_indices      |
| -        | 26.4±0.2ms                 | 5.98±0.07ms                            |    0.23 | frame_methods.Update.time_to_update_small_frame_big_arg |

**Review by rhshadrach — APPROVED** [review]
lgtm

**Comment by rhshadrach** [comment]
Thanks @aureliobarbosa!

**Pull request merged**